### PR TITLE
feat(widget): use flexible config for partner fee

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Docs update: https://github.com/cowprotocol/docs/pull/392

Since partner fee can be set for any trade type on any network, we should provide an option to flexibly configurate the fee.

Now both bps and recipient have type `FlexibleConfig`.

In advance, I've added `modules/volumeFee` because we are going to have fee in CoW Swap.
`volumeFee=cowswapFee | widgetPartnerFee`

# To Test

1. Use `window.cowSwapWidgetParams` in browser console (windget-configurator) tab to override partner fee.
2. Check arbitrary combinations of the configuration bellow:

```
window.cowSwapWidgetParams = {
    bps: 100,
    recipient: '0x0000000000000000000000000000000000000000'
}
```

```
window.cowSwapWidgetParams = {
    bps: {
        1: {
         'swap': 100, 'limit': 90, 'advanced': 80,
        },
        42161: {
            'swap': 70, 'limit': 60, 'advanced': 50,
        },
        100: {
            'swap': 40, 'limit': 30, 'advanced': 20,
        },
        11155111: {
            'swap': 10, 'limit': 5, 'advanced': 2,
        },
    },
    recipient: '0x0000000000000000000000000000000000000000',
}
```

```
window.cowSwapWidgetParams = {
    bps: 100,
    recipient: {
        'swap': {
            1: '0x...a', 42161: '0x...b', 100: '0x...c', 11155111: '0x...d',
        },
        'limit': {
            1: '0x...e', 42161: '0x...f', 100: '0x...g', 11155111: '0x...h',
        },
        'advanced': {
            1: '0x...j', 42161: '0x...i', 100: '0x...k', 11155111: '0x...l',
        },
    }
}
```

### CODE REVIEW

**Review by anxolin — COMMENTED** [review]
I really like the flexible config, and the fact that is not only for the partner fee but we can use it for any other piece of config. I added a small concern I have with the current implementation though

The config options you give (API) looks great to me. 

One thing I didn't fully get is... why changing the partnerFee to `volumeFee`, if this is because CoW Swap will start using it, I don't fully get it, that would be just make us have a default `partnerFee` for non-widget mode.

**Comment by shoom3301** [comment]
>One thing I didn't fully get is... why changing the partnerFee to volumeFee, if this is because CoW Swap will start using it, I don't fully get it, that would be just make us have a default partnerFee for non-widget mode.

@anxolin another reason why I separated them is that `partnerFee` has `FlexibleConfig` type and `volumeFee` has `{ bps: number, recipient: string }`. Using `resolveFlexibleConfig` we turn `partnerFee` into `volumeFee`

File: libs/widget-lib/src/flexibleConfig.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  return config as T
+}
+
+export function isPerTradeTypeConfig<T>(config: FlexibleConfig<T>): config is PerTradeTypeConfig<T> {
+  if (typeof config !== 'object') return false
+
+  return Object.keys(config as object).every((key) => TradeTypes.includes(key as TradeType))
+}
+
+export function isPerNetworkConfig<T>(config: FlexibleConfig<T>): config is PerNetworkConfig<T> {
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply libs/widget-lib/src/flexibleConfig.ts:33]: >but this is assuming the configuration for flows and networks is complete

No, it doesn't. You can set config only for some networks or flows:

```
{
    partnerFee: {
        bps: {
        1: {
         'swap': 100
        },
        100: {
            'limit': 30
        },
    },
     recipient: '0x9FA3c00a92Ec5f96B1Ad2527ab41B3932EFEDa58'
    }
}
```

[the remark this answers is not in the document]
Comment (shoom3301) [reply libs/widget-lib/src/flexibleConfig.ts:33]: Oooh, sorry, I've confused you in types.
Fixed: https://github.com/cowprotocol/cowswap/pull/4546/commits/b2474e3317b69fadf9efea6b6032b132034cb23c

**Review by elena-zh — APPROVED** [review]

File: apps/cowswap-frontend/src/modules/injectedWidget/utils/validatePartnerFee.test.ts [hunk 2]
```diff
@@ -54,6 +55,94 @@ describe('validatePartnerFee()', () => {
     })
   })
 
+  describe('When bps is a map', () => {
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/injectedWidget/utils/validatePartnerFee.test.ts:58]: Can you add a test for when both bps and recipient are maps?

Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/injectedWidget/utils/validatePartnerFee.test.ts:58]: image

There are both of them

Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/injectedWidget/utils/validatePartnerFee.test.ts:58]: Thanks, added

File: libs/widget-lib/src/flexibleConfig.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (anxolin) [reply libs/widget-lib/src/flexibleConfig.ts:33]: Awesome!

**Review by anxolin — APPROVED** [review]

**Pull request merged**